scipy/optimize/_hessian_update_strategy.py

```diff
@@ -0,0 +1,338 @@
+"""quasi-Newton update strategies."""
+
+from __future__ import division, print_function, absolute_import
+import numpy as np
+from numpy.linalg import norm
+from scipy.linalg import get_blas_funcs
+
+class HessianUpdateStrategy(object):
+    """Virtual interface for implementing Hessian update strategies.
```

Comment:
Also explain a little bit more what this approximation means and what it stores inside (like an approximation of the Hessian or its inverse). I suggest to introduce a term "internal matrix" here.

Reply:
Ok

Reply:
I wrote a additional paragraph, trying to clarify the purpose of this interface:

```
    Many optimization methods make use of Hessian (or inverse Hessian)
    quasi-Newton approximations, such as BFGS, SR1, L-BFGS. Some of these
    approximations, however, do not actually need to store the entire matrix
    or can compute the  internal matrix product with a given vector in a very
    efficiently manner. This class serves as an abstract interface between
    the optimization algorithm and the quasi-Newton update strategies, giving
    freedom of implementation to store and update the internal matrix as
    efficiently as possible. Different choices of initialization and update
    procedure will result in different quasi-Newton strategies.
```

Reply:
Looks excellent.

Reply:
One thing I'm not convinced is "quasi-Newton approximation". There are "quasi-Newton methods" which use Hessian approximations, but "quasi-Newton approximations" is not widely used.

For example, in this case we can write "Many optimization methods make use of Hessian (or inverse Hessian) approximations (so called quasi-Newton methods)". I think it will match book style texts better.

If you decide to apply this change, then I suggest to remove "quasi-Newton approximation" phrase from everywhere (rephrase or simply use "Hessian approximation" instead).

Reply:
Ok I agree with you. I will remove "quasi-Newton approximation"  from everywhere.